ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -197,27 +157,39 @@ public void onError(Throwable t) {
                     t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
                     t
                 );
-            response(tensorOutputs);
+            response();
         }
 
         @Override
         public void onComplete() {
-            response(tensorOutputs);
+            response();
         }
     }
 
-    private void response(Map<Integer, ModelTensors> tensors) {
-        processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
-        executionContext.getCountDownLatch().countDown();
-        // when countdown's count equals to 0 means all responses are received.
-        if (executionContext.getCountDownLatch().getCount() == 0) {
-            if (executionContext.getExceptionHolder().get() != null) {
-                actionListener.onFailure(executionContext.getExceptionHolder().get());
-                return;
-            }
-            reOrderTensorResponses(tensors);
+    private void response() {
+        String body = responseBody.toString();
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            handleException(new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST), null);
         } else {
-            log.debug("Not all responses received, left response count is: " + executionContext.getCountDownLatch().getCount());
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                handleException(
+                    new OpenSearchStatusException(REMOTE_SERVICE_ERROR + body, RestStatus.fromCode(statusCode)),
+                    String.format(Locale.ROOT, "Remote server returned error code: %d", statusCode)
```

Comment:
It seems only here is passing the errorMsg, can add log here and change the handleException method to accept only one parameter, cause a lot of invocation passing `null` doesn't seem neat. Also I prefer to move the actionListener.onFailure to handleException method, so this method will be more clean.

Reply:
We use the exceptionHolder because @Zhangxunmt call out that we may parse more errors except throttling from onHeaders. Since we need to cache the exceptions and call onFailure in response() method, using exceptionHolder can reduces boolean flags.
I just realized that we can only use the exceptionHolder in onHeaders, and directly call onFailure at response(). This is a more neat way

Reply:
updated